src/preload.py

```diff
@@ -0,0 +1,329 @@
+    response = get(url, params=params, headers=headers)
+    response.raise_for_status()
+    return response.json()
+
+
+def get_registry_token(image_repo):
+    params = {
+        "service": REGISTRY_HOST,
+        "scope": "repository:{}:pull".format(image_repo),
+    }
+    return api("auth/v1/token", params)["token"]
+
+
+def registry(endpoint, registry_token, headers=None, json=True):
+    headers = headers or {}
+    headers["Authorization"] = "Bearer {}".format(registry_token)
+    url = "https://{}/{}".format(REGISTRY_HOST, endpoint)
+    response = get(url, headers=headers)
+    response.raise_for_status()
+    return response.json() if json else response.content
+
+
+def get_app_data():
+    # Fetch application metadata
+    endpoint = "v2/application({})?$expand=environment_variable".format(APP_ID)
+    data = api(endpoint)["d"][0]
+    image_repo = "{app_name}/{commit}".format(**data).lower()
+    image_id = "{}/{}".format(REGISTRY_HOST, image_repo).lower()
+    env = {v["name"]: v["value"] for v in data.get("environment_variable", [])}
+    return {
+        "appId": data["id"],
+        "name": data["app_name"],
+        "commit": data["commit"],
+        "imageRepo": image_repo,
+        "imageId": image_id,
+        "env": {k: v for k, v in env.items() if not k.startswith("RESIN_")},
+        "config": {k: v for k, v in env.items() if k.startswith("RESIN_")},
+    }
+
+
+def get_container_size(image_repo):
+    log("Fetching container size for {}".format(image_repo))
+    token = get_registry_token(image_repo)
+    data = registry("v2/{}/manifests/latest".format(image_repo), token)
+    headers = {"Range": "bytes=-4"}
+    size = 0
+    # the last 4 bytes of each gzipped layer are the layer size % 32
+    for layer in data["fsLayers"]:
+        endpoint = "v2/{}/blobs/{}".format(image_repo, layer["blobSum"])
+        layer_size_bytes = registry(endpoint, token, headers, json=False)
+        layer_size = int.from_bytes(layer_size_bytes, byteorder="little")
+        size += layer_size
+        log(endpoint, size)
+    log("Container size: {}".format(human_size(size)))
+    return size
+
+
+def get_resin_os_version(image):
+    with mount_context_manager(image, 2) as mountpoint:
+        with open(os.path.join(mountpoint, "etc/os-release")) as f:
+            for line in f:
+                key, value = line.split("=", 1)
+                if key == "VERSION":
+                    return value.strip('"\n"')
+
+
+def expand_image(image, additional_space):
+    log("Expanding image size by", human_size(additional_space))
+    # Add zero bytes to image to be able to resize partitions
+    with open(image, "a") as f:
+        size = f.tell()
+        f.truncate(size + additional_space)
+
+
+def expand_partitions(image):
+    # Resizes partition 4 & 6 to the end of the image
+    log("Expanding extended partition 4 to the end of the disk image.")
+    log("Expanding logical partition 6 to the end of the disk image.")
```

Comment:
Since below it's a single operation, it would be good to have a single line of log as well (multiple log lines to me mean multiple tasks, and things happening between log lines)